File: tools/meshing.py
#!/usr/bin/env python3
"""Turn a scanned point cloud into a triangle mesh.

Three reconstructions, all of them standard:

  poisson   Screened Poisson surface reconstruction (Kazhdan). Watertight,
            smooth, tolerant of noise -- the usual default for scanner data.
            Needs oriented normals, and needs open3d.
  bpa       Ball pivoting. Keeps the sampled points as mesh vertices, so it
            preserves detail and leaves genuine holes as holes. Needs open3d.
  alpha     3D alpha shape off a Delaunay tetrahedralisation. Cruder than the
            other two, but it is the one that runs on a plain scipy install,
            so there is always something available.

Poisson and ball pivoting each come from whichever of two libraries is
present. They are the same published algorithms either way:

  pymeshlab  MeshLab's build. Has wheels for current Pythons, including 3.13,
             which is the only reason this project can offer Poisson at all --
             open3d stops at 3.12.
  open3d     The other common packaging, for older interpreters.

Both are optional. `available_methods()` reports what this machine can
actually do, and the UI only offers those, rather than failing at the point
where the user presses Build.

Everything here works in millimetres, matching the rest of the pipeline.
"""

import numpy as np

try:
    import pymeshlab as ml
    HAVE_PYMESHLAB = True
except Exception:
    ml = None
    HAVE_PYMESHLAB = False

try:
    import open3d as o3d
    HAVE_OPEN3D = True
except Exception:      # not installed, or installed and broken -- same to us
    o3d = None
    HAVE_OPEN3D = False

# pymeshlab first: it is the one that exists on the Pythons this runs on.
HAVE_RECON = HAVE_PYMESHLAB or HAVE_OPEN3D


# name -> (label, needs a reconstruction backend)
METHODS = {
    "poisson": ("Poisson", True),
    "bpa": ("Ball pivoting", True),
    "alpha": ("Alpha shape", False),
}

# How many points each method is worth feeding, measured on this project's
# clouds. They differ by nearly an order of magnitude because the algorithms
# scale so differently:
#
#   poisson  roughly linear -- 180k in about 13 s. It also fits a smooth
#            surface rather than interpolating points, so extra samples past
#            what the octree resolves buy nothing anyway.
#   bpa      badly superlinear -- 45k in 8 s but 120k in 129 s, so a 270k
#            cloud is twenty minutes. This is the one that needs a hard cap.
#   alpha    Delaunay is about six tets per point, and memory rather than
#            time is the wall.
METHOD_BUDGET = {"poisson": 150000, "bpa": 40000, "alpha": 80000}


def backend():
    """Which library is doing the reconstruction, for the UI to report."""
    if HAVE_PYMESHLAB:
        return "pymeshlab"
    if HAVE_OPEN3D:
        return "open3d"
    return None


def available_methods():
    """Method keys usable on this machine, best first."""
    return [k for k, (_, needs) in METHODS.items() if HAVE_RECON or not needs]


def missing_note():
    """Why the good methods are absent, or None when they are not."""
    if HAVE_RECON:
        return None
    return ("no reconstruction backend -- only the alpha shape is available.\n"
            "  pip install pymeshlab   for Poisson and ball pivoting.\n"
            "  (open3d also works, but has no wheels past Python 3.12.)")


# --- shared helpers ---------------------------------------------------------

def spacing(xyz, k=1):
    """Median distance to the k-th nearest neighbour. The natural length unit
    of a cloud: every radius below is quoted as a multiple of it, so one set of
    defaults works for a desk and for a room."""
    from scipy.spatial import cKDTree
    if len(xyz) < 2:
        return 1.0
    d, _ = cKDTree(xyz).query(xyz, k=min(k + 1, len(xyz)))
    return max(float(np.median(d[:, -1])), 1e-6)


def voxel_downsample(xyz, size):
    """One point per occupied cell of a `size` grid, at the cell's centroid.

    Averaging rather than picking a representative: the mean of the points in
    a cell is a better estimate of the surface than any one of them, so this
    denoises slightly as it thins.
    """
    if size <= 0:
        return xyz
    key = np.floor(xyz / size).astype(np.int64)
    _, inv = np.unique(key, axis=0, return_inverse=True)
    inv = inv.ravel()
    n = inv.max() + 1
    out = np.zeros((n, 3))
    np.add.at(out, inv, xyz)
    return out / np.bincount(inv, minlength=n)[:, None]


def downsample_to_budget(xyz, budget):
    """Thin to about `budget` points on a uniform grid.

    Uniform is the operative word. A lidar sweep is wildly non-uniform -- dense
    near the scanner and along the spin axis, sparse far away -- and both the
    alpha shape and ball pivoting take a single radius for the whole cloud, so
    non-uniform density is precisely what makes them fail: any radius large
    enough to close the sparse regions has already swallowed the detail in the
    dense ones. That is the "holes or no detail, pick one" trade, and it is a
    property of the sampling rather than of the algorithms.

    Thinning on a grid removes the variation instead of preserving it the way
    random subsampling does, so one radius then works everywhere.
    """
    if budget <= 0 or len(xyz) <= budget:
        return xyz, 0.0

    # Bracket a voxel size that lands near the budget, then bisect. Point count
    # falls monotonically with cell size, so this converges quickly, and it is
    # far cheaper than it looks: each probe is one pass of integer division.
    lo, hi = 0.0, float(np.ptp(xyz, axis=0).max())
    span = hi
    best = xyz
    for _ in range(24):
        mid = 0.5 * (lo + hi)
        if mid <= 0 or mid < span * 1e-6:
            break
        out = voxel_downsample(xyz, mid)
        if len(out) > budget:
            lo = mid
        else:
            best, hi = out, mid
            # Close enough: chasing the exact figure costs passes and buys
            # nothing, since the budget is a rough guard rail anyway.
            if len(out) > budget * 0.9:
                break
    return best, hi


def _to_o3d(xyz, normal_radius, normal_k):
    """Point cloud with normals oriented towards the scanner.

    Normal *orientation* is the part of Poisson that usually goes wrong, and
    it is free here: the lidar sits at the world origin, so every surface it
    saw is a surface facing it. That beats the generic minimum-spanning-tree
    propagation, which flips whole patches on scenes with thin structure.
    """
    pcd = o3d.geometry.PointCloud()
    pcd.points = o3d.utility.Vector3dVector(np.asarray(xyz, dtype=np.float64))
    pcd.estimate_normals(
        search_param=o3d.geometry.KDTreeSearchParamHybrid(
            radius=normal_radius, max_nn=normal_k))
    pcd.orient_normals_towards_camera_location(np.zeros(3))
    return pcd


def _from_o3d(mesh):
    return (np.asarray(mesh.vertices, dtype=np.float32),
            np.asarray(mesh.triangles, dtype=np.int32))


# --- the three reconstructions ----------------------------------------------

def _pml_cloud(xyz, normal_k):
    """MeshLab point cloud with normals pointed back at the scanner.

    Same trick as the open3d path: viewpos is the world origin, which is
    where the lidar physically is, so "facing the viewpoint" is exactly right
    rather than a heuristic.
    """
    ms = ml.MeshSet()
    ms.add_mesh(ml.Mesh(vertex_matrix=np.asarray(xyz, dtype=np.float64)))
    ms.compute_normal_for_point_clouds(
        k=int(normal_k), smoothiter=0, flipflag=True,
        viewpos=np.zeros(3))
    return ms


def _from_pml(ms):
    m = ms.current_mesh()
    return (np.asarray(m.vertex_matrix(), dtype=np.float32),
            np.asarray(m.face_matrix(), dtype=np.int32))


def _poisson(xyz, depth, normal_radius, normal_k):
    if HAVE_PYMESHLAB:
        ms = _pml_cloud(xyz, normal_k)
        # preclean drops the duplicate and unreferenced vertices that a lidar
        # cloud is full of, which the solver would otherwise weight twice.
        ms.generate_surface_reconstruction_screened_poisson(
            depth=int(depth), preclean=True)
        return _from_pml(ms)

    pcd = _to_o3d(xyz, normal_radius, normal_k)
    mesh, _density = o3d.geometry.TriangleMesh.create_from_point_cloud_poisson(
        pcd, depth=int(depth))
    return _from_o3d(mesh)


def _bpa(xyz, radius_mult, normal_radius, normal_k):
    r = spacing(np.asarray(xyz)) * float(radius_mult)
    if HAVE_PYMESHLAB:
        ms = _pml_cloud(xyz, normal_k)
        # AbsoluteValue was renamed PureValue somewhere around the 2023
        # releases, and both spellings are still in the wild.
        absolute = getattr(ml, "PureValue", None) or ml.AbsoluteValue
        ms.generate_surface_reconstruction_ball_pivoting(
            ballradius=absolute(float(r)))
        return _from_pml(ms)

    pcd = _to_o3d(xyz, normal_radius, normal_k)
    # A ladder of radii, not one: a single ball either bridges the sparse
    # regions or shrink-wraps the dense ones, never both. Successive passes
    # with a bigger ball fill what the previous pass left open.
    radii = o3d.utility.DoubleVector([r, r * 2.0, r * 4.0])
    mesh = o3d.geometry.TriangleMesh.create_from_point_cloud_ball_pivoting(
        pcd, radii)
    return _from_o3d(mesh)


def trim_to_samples(verts, faces, xyz, max_dist):
    """Drop the parts of a surface that no measurement supports.

    Poisson always closes its surface, so it invents geometry across
    everything the lidar never saw -- scan one wall and it comes back wrapped
    round into a blob. The invented parts are exactly the parts far from any
    real sample, so this is what turns the blob back into a wall.

    Done here by distance rather than by the solver's own density estimate so
    that it means the same thing whichever backend ran, and so the threshold
    is a length the user can reason about.
    """
    if max_dist <= 0 or not len(faces):
        return verts, faces
    from scipy.spatial import cKDTree
    d, _ = cKDTree(np.asarray(xyz)).query(verts)
    keep = d <= max_dist
    if keep.all():
        return verts, faces
    # A triangle survives only with all three corners supported; dropping on
    # "any" would leave spikes reaching out to the trimmed-away region.
    faces = faces[keep[faces].all(axis=1)]
    if not len(faces):
        return verts, faces
    used, faces = np.unique(faces, return_inverse=True)
    return verts[used], faces.reshape(-1, 3).astype(np.int32)


def _alpha_scipy(xyz, alpha):
    """Boundary of the alpha complex, via Delaunay. numpy + scipy only.

    A tetrahedron is in the complex when its circumsphere is smaller than
    alpha; the surface is every triangle that ended up in exactly one such
    tetrahedron. Large alpha tends to the convex hull, small alpha crumbles
    into nothing, and the useful range is a few times the point spacing.
    """
    from scipy.spatial import Delaunay
    xyz = np.asarray(xyz, dtype=np.float64)
    tets = Delaunay(xyz).simplices
    if not len(tets):
        return np.zeros((0, 3), np.float32), np.zeros((0, 3), np.int32)

    # Circumradius of every tet, batched. The circumcentre solves
    # 2(p_i - p_0).c = |p_i|^2 - |p_0|^2 for i = 1..3.
    p = xyz[tets]                                     # (T, 4, 3)
    a = p[:, 0]
    A = 2.0 * (p[:, 1:] - a[:, None, :])              # (T, 3, 3)
    sq = (p ** 2).sum(axis=-1)                        # (T, 4)
    rhs = sq[:, 1:] - sq[:, :1]                       # (T, 3)

    # Flat (near-degenerate) tets have no usable circumcentre and would make
    # the batched solve raise for the whole array, so they are scored infinite
    # and simply never make the cut.
    scale = np.abs(A).max(axis=(1, 2))
    det = np.abs(np.linalg.det(A))
    ok = det > 1e-10 * np.maximum(scale, 1e-9) ** 3
    R = np.full(len(tets), np.inf)
    if ok.any():
        c = np.linalg.solve(A[ok], rhs[ok][..., None])[..., 0]
        R[ok] = np.linalg.norm(c - a[ok], axis=1)

    keep = tets[R < alpha]
    if not len(keep):
        return np.zeros((0, 3), np.float32), np.zeros((0, 3), np.int32)

    # Four faces per kept tet, each tagged with the vertex it faces away from.
    faces = np.concatenate([keep[:, [1, 2, 3]], keep[:, [0, 2, 3]],
                            keep[:, [0, 1, 3]], keep[:, [0, 1, 2]]])
    opp = np.concatenate([keep[:, 0], keep[:, 1], keep[:, 2], keep[:, 3]])

    # A face shared by two kept tets is interior; the surface is the rest.
    key = np.sort(faces, axis=1)
    _, inv, cnt = np.unique(key, axis=0, return_inverse=True,
                            return_counts=True)
    surf = cnt[inv.ravel()] == 1
    faces, opp = faces[surf], opp[surf]

    # Point every triangle away from the solid, so lit shading reads as a
    # surface instead of a patchwork of dark and bright facets.
    v0, v1, v2 = xyz[faces[:, 0]], xyz[faces[:, 1]], xyz[faces[:, 2]]
    n = np.cross(v1 - v0, v2 - v0)
    inward = np.einsum("ij,ij->i", n, xyz[opp] - v0) > 0
    faces[inward] = faces[inward][:, [0, 2, 1]]

    used, faces = np.unique(faces, return_inverse=True)
    return (xyz[used].astype(np.float32),
            faces.reshape(-1, 3).astype(np.int32))


def _alpha(xyz, alpha):
    # No pymeshlab branch: MeshLab has no alpha-shape filter, and the scipy
    # implementation below is the point of this method anyway.
    if HAVE_OPEN3D:
        pcd = o3d.geometry.PointCloud()
        pcd.points = o3d.utility.Vector3dVector(np.asarray(xyz, np.float64))
        mesh = o3d.geometry.TriangleMesh.create_from_point_cloud_alpha_shape(
            pcd, float(alpha))
        return _from_o3d(mesh)
    return _alpha_scipy(xyz, alpha)


# --- smoothing --------------------------------------------------------------

def smooth(verts, faces, iters, lam=0.5):
    """Laplacian smoothing: nudge each vertex towards its neighbours' mean.

    Mainly for the alpha shape, whose surface is made of sampled points and so
    carries the lidar's per-sample noise as visible faceting.
    """
    if iters <= 0 or not len(faces):
        return verts
    e = np.concatenate([faces[:, [0, 1]], faces[:, [1, 2]], faces[:, [2, 0]]])
    e = np.concatenate([e, e[:, ::-1]])          # both directions
    deg = np.bincount(e[:, 0], minlength=len(verts)).astype(np.float32)
    deg[deg == 0] = 1.0
    v = verts.astype(np.float32).copy()
    for _ in range(int(iters)):
        acc = np.zeros_like(v)
        np.add.at(acc, e[:, 0], v[e[:, 1]])
        v += lam * (acc / deg[:, None] - v)
    return v


# --- entry point ------------------------------------------------------------

def reconstruct(xyz, method="poisson", *, depth=9, trim_mult=3.0,
                radius_mult=2.0, alpha_mult=4.0, normal_k=30,
                normal_mult=4.0, smooth_iters=0, budget=0):
    """Mesh a cloud. Returns (verts (V,3) float32, faces (F,3) int32, info).

    The radius-ish parameters are multiples of the cloud's own point spacing,
    not absolute millimetres, so the defaults survive a change of subject.

    `budget` thins the input to about that many points on a uniform grid
    first; 0 means feed the lot. info reports what actually happened, since
    the thinning is otherwise invisible and it changes the result.

    Raises ValueError with something worth reading when a method is not
    available or the cloud is too small to mesh.
    """
    xyz = np.asarray(xyz, dtype=np.float64)
    xyz = xyz[np.isfinite(xyz).all(axis=1)]
    if len(xyz) < 4:
        raise ValueError("need at least 4 points to build a surface")
    if method not in METHODS:
        raise ValueError(f"unknown method {method!r}")
    if METHODS[method][1] and not HAVE_RECON:
        raise ValueError(f"{METHODS[method][0]} needs pymeshlab "
                         f"(pip install pymeshlab)")

    info = {"input": len(xyz), "voxel": 0.0}
    if budget:
        xyz, info["voxel"] = downsample_to_budget(xyz, int(budget))
    info["used"] = len(xyz)
    if len(xyz) < 4:
        raise ValueError("too few points left after thinning")

    s = spacing(xyz)
    info["spacing"] = s
    normal_radius = s * float(normal_mult)

    if method == "poisson":
        verts, faces = _poisson(xyz, depth, normal_radius, normal_k)
        verts, faces = trim_to_samples(verts, faces, xyz,
                                       s * float(trim_mult))
    elif method == "bpa":
        verts, faces = _bpa(xyz, radius_mult, normal_radius, normal_k)
    else:
        verts, faces = _alpha(xyz, s * float(alpha_mult))

    if smooth_iters and len(faces):
        verts = smooth(verts, faces, smooth_iters)
    return verts.astype(np.float32), faces.astype(np.int32), info


# --- export -----------------------------------------------------------------

def export_ply(path, verts, faces, rgb=None):
    """Binary PLY with faces. rgb is (V,3) uint8, or None."""
    props = [("x", "<f4"), ("y", "<f4"), ("z", "<f4")]
    if rgb is not None:
        props += [("red", "u1"), ("green", "u1"), ("blue", "u1")]
    vert = np.empty(len(verts), dtype=props)
    vert["x"], vert["y"], vert["z"] = verts[:, 0], verts[:, 1], verts[:, 2]
    if rgb is not None:
        vert["red"], vert["green"], vert["blue"] = rgb[:, 0], rgb[:, 1], rgb[:, 2]

    # PLY face lists are per-element counted, so the count column rides along
    # in the same record rather than being written separately.
    face = np.empty(len(faces), dtype=[("n", "u1"), ("v", "<i4", 3)])
    face["n"] = 3
    face["v"] = faces

    with open(path, "wb") as fh:
        fh.write(b"ply\nformat binary_little_endian 1.0\n")
        fh.write(f"element vertex {len(verts)}\n".encode())
        fh.write(b"property float x\nproperty float y\nproperty float z\n")
        if rgb is not None:
            fh.write(b"property uchar red\nproperty uchar green\n"
                     b"property uchar blue\n")
        fh.write(f"element face {len(faces)}\n".encode())
        fh.write(b"property list uchar int vertex_indices\n")
        fh.write(b"end_header\n")
        fh.write(vert.tobytes())
        fh.write(face.tobytes())


def export_stl(path, verts, faces):
    """Binary STL. No colour and no shared vertices, but every slicer and CAD
    package reads it, which is the point of offering it at all."""
    tri = verts[faces].astype(np.float32)                  # (F, 3, 3)
    n = np.cross(tri[:, 1] - tri[:, 0], tri[:, 2] - tri[:, 0])
    ln = np.linalg.norm(n, axis=1, keepdims=True)
    n = np.divide(n, ln, out=np.zeros_like(n), where=ln > 0)

    rec = np.zeros(len(faces), dtype=[("n", "<f4", 3), ("v", "<f4", (3, 3)),
                                      ("attr", "<u2")])
    rec["n"], rec["v"] = n, tri
    with open(path, "wb") as fh:
        fh.write(b"\0" * 80)
        fh.write(np.uint32(len(faces)).tobytes())
        fh.write(rec.tobytes())

File: app/src-tauri/src/meshing.rs
//! Point cloud -> triangle mesh, with screened Poisson reconstruction.
//!
//! The pipeline: thin to a point budget on a uniform grid (a lidar sweep is
//! wildly non-uniform), estimate normals by local PCA and orient each one
//! towards the scanner that saw the point -- orientation is the part of
//! Poisson that usually goes wrong, and here it is known exactly -- solve,
//! then trim away the surface no measurement supports (Poisson closes
//! everything), drop the crumbs that trimming leaves, and optionally smooth.

use rayon::prelude::*;
use serde::Serialize;

use crate::kdtree::KdTree;
use crate::poisson::{self, PrParams};
use crate::util::{self, Stopwatch, V3};

#[derive(Clone, Debug)]
pub struct MeshParams {
    /// Finest surface detail wanted, mm. The octree depth follows from it
    /// and the size of the scene, so a desk and a building get the same
    /// physical resolution.
    pub cell_mm: f64,
    pub samples_per_node: f32,
    pub point_weight: f32,
    /// Drop surface further than this many point spacings from any sample.
    pub trim: f64,
    /// Thin the input to about this many points first (0 = no limit).
    pub budget: usize,
    pub smooth_iters: u32,
    pub normal_k: usize,
}

impl MeshParams {
    /// Quality presets: "medium", "high" (the default), "max".
    pub fn preset(quality: &str, trim: f64, budget: usize, smooth: u32) -> MeshParams {
        let (cell_mm, spn) = match quality {
            "medium" => (10.0, 2.0),
            "max" => (2.5, 1.5),
            _ => (5.0, 1.5),
        };
        MeshParams {
            cell_mm,
            samples_per_node: spn,
            point_weight: 2.0,
            trim,
            budget,
            smooth_iters: smooth,
            normal_k: 24,
        }
    }
}

pub struct Mesh {
    pub verts: Vec<[f32; 3]>,
    pub normals: Vec<[f32; 3]>,
    pub tris: Vec<[u32; 3]>,
}

#[derive(Clone, Debug, Default, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MeshInfo {
    pub input: usize,
    pub used: usize,
    pub voxel: f64,
    pub spacing: f64,
    pub depth: u32,
    pub cell_mm: f64,
    pub verts: usize,
    pub tris: usize,
    pub seconds: f64,
}

#[derive(Clone, Copy, Debug, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum MeshStage {
    Thinning,
    Normals,
    Solving,
    Cleaning,
}

/// Centroid per cell of a `size` grid; each output point keeps the scanner
/// index of the first point in its cell.
fn thin(p: &[V3], scan: &[u16], size: f64) -> (Vec<V3>, Vec<u16>) {
    let (cell, n, _) = util::group_cells(p, size, 0.0);
    let mut sum = vec![[0.0f64; 4]; n];
    let mut sc = vec![u16::MAX; n];
    for (i, q) in p.iter().enumerate() {
        let c = cell[i] as usize;
        let s = &mut sum[c];
        s[0] += q[0];
        s[1] += q[1];
        s[2] += q[2];
        s[3] += 1.0;
        if sc[c] == u16::MAX {
            sc[c] = scan[i];
        }
    }
    (sum.into_iter().map(|s| [s[0] / s[3], s[1] / s[3], s[2] / s[3]]).collect(), sc)
}

fn extent(p: &[V3]) -> f64 {
    let mut mn = [f64::INFINITY; 3];
    let mut mx = [f64::NEG_INFINITY; 3];
    for q in p {
        for d in 0..3 {
            mn[d] = mn[d].min(q[d]);
            mx[d] = mx[d].max(q[d]);
        }
    }
    (mx[0] - mn[0]).max(mx[1] - mn[1]).max(mx[2] - mn[2]).max(1.0)
}

/// Thins the cloud for a solve at `depth`.
///
/// A lidar sweep is wildly anisotropic: consecutive slices of a slow sweep sit
/// a millimetre or two apart while points along a slice are centimetres
/// apart, so the raw cloud is mostly redundancy the octree merges anyway. It
/// is thinned to a grid just finer than the finest octree cell -- nothing
/// below that can show in the surface -- which also makes the spacing, and so
/// the trim distance, mean what it says. The point budget caps what is left.
fn downsample_for_depth(p: &[V3], scan: &[u16], depth: u32, budget: usize) -> (Vec<V3>, Vec<u16>, f64) {
    let finest = 1.1 * extent(p) / (1u64 << depth) as f64;
    let mut size = 0.75 * finest;
    let (mut out, mut sc) = thin(p, scan, size);
    if budget > 0 && out.len() > budget {
        // Grow the cell until the count fits: count falls roughly with the
        // square of the cell on surfaces.
        for _ in 0..12 {
            size *= ((out.len() as f64 / budget as f64).sqrt() * 1.02).max(1.02);
            let t = thin(p, scan, size);
            out = t.0;
            sc = t.1;
            if out.len() <= budget {
                break;
            }
        }
    }
    (out, sc, size)
}

/// Median distance to the nearest neighbour, from a subsample.
fn spacing(p: &[V3], tree: &KdTree) -> f64 {
    if p.len() < 2 {
        return 1.0;
    }
    let step = (p.len() / 100_000).max(1);
    let mut d: Vec<f64> = (0..p.len())
        .into_par_iter()
        .step_by(step)
        .map_init(Vec::new, |buf, i| {
            tree.knn(p[i], 2, buf);
            buf.get(1).map(|x| x.1.sqrt()).unwrap_or(0.0)
        })
        .collect();
    util::median_mut(&mut d).max(1e-6)
}

/// PCA normals, flipped to face the scanner that measured each point.
fn oriented_normals(p: &[V3], scan: &[u16], origins: &[V3], tree: &KdTree, k: usize) -> Vec<V3> {
    let k = k.min(p.len());
    p.par_iter()
        .enumerate()
        .map_init(Vec::new, |buf, (i, &q)| {
            tree.knn(q, k, buf);
            let (_, cov, n) = util::centered_cov(buf.iter().map(|&(j, _)| p[j]));
            let mut nn = if n >= 3 { util::sym3_eigen(cov).1[0] } else { [0.0, 0.0, 1.0] };
            let o = origins.get(scan[i] as usize).copied().unwrap_or([0.0; 3]);
            if util::dot(nn, util::sub(o, q)) < 0.0 {
                nn = util::scale(nn, -1.0);
            }
            nn
        })
        .collect()
}

/// Keeps triangles whose three corners are within `max_dist` of a sample, then
/// compacts the vertex array.
fn trim_to_samples(verts: Vec<[f32; 3]>, tris: Vec<[u32; 3]>, tree: &KdTree, max_dist: f64) -> (Vec<[f32; 3]>, Vec<[u32; 3]>) {
    if max_dist <= 0.0 || tris.is_empty() {
        return (verts, tris);
    }
    let md2 = max_dist * max_dist;
    let keep: Vec<bool> = verts
        .par_iter()
        .map(|v| tree.nearest([v[0] as f64, v[1] as f64, v[2] as f64], md2).is_some())
        .collect();
    let tris: Vec<[u32; 3]> = tris.into_par_iter().filter(|t| keep[t[0] as usize] && keep[t[1] as usize] && keep[t[2] as usize]).collect();
    compact(verts, tris)
}

fn compact(verts: Vec<[f32; 3]>, tris: Vec<[u32; 3]>) -> (Vec<[f32; 3]>, Vec<[u32; 3]>) {
    let mut map = vec![u32::MAX; verts.len()];
    let mut out = Vec::new();
    let mut nt = Vec::with_capacity(tris.len());
    for t in tris {
        let mut r = [0u32; 3];
        for k in 0..3 {
            let v = t[k] as usize;
            if map[v] == u32::MAX {
                map[v] = out.len() as u32;
                out.push(verts[v]);
            }
            r[k] = map[v];
        }
        nt.push(r);
    }
    (out, nt)
}

fn find(parent: &mut [u32], mut x: u32) -> u32 {
    while parent[x as usize] != x {
        let p = parent[x as usize];
        parent[x as usize] = parent[p as usize];
        x = p;
    }
    x
}

/// Removes connected pieces that are tiny next to the whole: the crumbs
/// trimming leaves along the fringe of the scanned surface.
fn drop_crumbs(verts: Vec<[f32; 3]>, tris: Vec<[u32; 3]>) -> (Vec<[f32; 3]>, Vec<[u32; 3]>) {
    if tris.is_empty() {
        return (verts, tris);
    }
    let mut parent: Vec<u32> = (0..verts.len() as u32).collect();
    for t in &tris {
        let a = find(&mut parent, t[0]);
        for k in 1..3 {
            let b = find(&mut parent, t[k]);
            if a != b {
                parent[b as usize] = a;
            }
        }
    }
    let mut faces = std::collections::HashMap::new();
    let roots: Vec<u32> = tris.iter().map(|t| find(&mut parent, t[0])).collect();
    for &r in &roots {
        *faces.entry(r).or_insert(0usize) += 1;
    }
    let min_faces = (tris.len() / 2000).min(500).max(24);
    let tris: Vec<[u32; 3]> = tris.into_iter().zip(roots).filter(|(_, r)| faces[r] >= min_faces).map(|(t, _)| t).collect();
    compact(verts, tris)
}

fn laplacian_smooth(verts: &mut [[f32; 3]], tris: &[[u32; 3]], iters: u32) {
    if iters == 0 || tris.is_empty() {
        return;
    }
    let n = verts.len();
    let mut deg = vec![0f32; n];
    for t in tris {
        for k in 0..3 {
            deg[t[k] as usize] += 2.0;
        }
    }
    for _ in 0..iters {
        let mut acc = vec![[0f32; 3]; n];
        for t in tris {
            for k in 0..3 {
                let a = t[k] as usize;
                for m in [(k + 1) % 3, (k + 2) % 3] {
                    let b = t[m] as usize;
                    for d in 0..3 {
                        acc[a][d] += verts[b][d];
                    }
                }
            }
        }
        for i in 0..n {
            if deg[i] > 0.0 {
                for d in 0..3 {
                    verts[i][d] += 0.5 * (acc[i][d] / deg[i] - verts[i][d]);
                }
            }
        }
    }
}

/// Area-weighted vertex normals.
pub fn vertex_normals(verts: &[[f32; 3]], tris: &[[u32; 3]]) -> Vec<[f32; 3]> {
    let mut acc = vec![[0f64; 3]; verts.len()];
    for t in tris {
        let a = verts[t[0] as usize];
        let b = verts[t[1] as usize];
        let c = verts[t[2] as usize];
        let u = [(b[0] - a[0]) as f64, (b[1] - a[1]) as f64, (b[2] - a[2]) as f64];
        let v = [(c[0] - a[0]) as f64, (c[1] - a[1]) as f64, (c[2] - a[2]) as f64];
        let n = util::cross(u, v);
        for k in 0..3 {
            let s = &mut acc[t[k] as usize];
            s[0] += n[0];
            s[1] += n[1];
            s[2] += n[2];
        }
    }
    acc.into_par_iter()
        .map(|n| {
            let l = util::norm(n);
            if l > 0.0 {
                [(n[0] / l) as f32, (n[1] / l) as f32, (n[2] / l) as f32]
            } else {
                [0.0, 0.0, 1.0]
            }
        })
        .collect()
}

/// Meshes `points` (world frame). `scan` gives each point's scanner as an
/// index into `origins` (the scanner positions), for normal orientation.
pub fn reconstruct(
    points: &[V3],
    scan: &[u16],
    origins: &[V3],
    prm: &MeshParams,
    progress: &(dyn Fn(MeshStage) + Sync),
    cancelled: &(dyn Fn() -> bool + Sync),
) -> Result<(Mesh, MeshInfo), String> {
    let sw = Stopwatch::start();
    let mut info = MeshInfo { input: points.len(), ..Default::default() };
    if points.len() < 100 {
        return Err("too few points".into());
    }
    progress(MeshStage::Thinning);
    let ext = extent(points);
    let depth = ((1.1 * ext / prm.cell_mm.max(0.5)).log2().round() as i64).clamp(7, 13) as u32;
    info.depth = depth;
    info.cell_mm = 1.1 * ext / (1u64 << depth) as f64;
    let (p, sc, voxel) = downsample_for_depth(points, scan, depth, prm.budget);
    info.used = p.len();
    info.voxel = voxel;
    if cancelled() {
        return Err("cancelled".into());
    }
    progress(MeshStage::Normals);
    let tree = KdTree::new(&p);
    let s = spacing(&p, &tree);
    info.spacing = s;
    let normals = oriented_normals(&p, &sc, origins, &tree, prm.normal_k);
    if cancelled() {
        return Err("cancelled".into());
    }
    progress(MeshStage::Solving);
    let pf: Vec<[f32; 3]> = util::to_f32(&p);
    let nf: Vec<[f32; 3]> = util::to_f32(&normals);
    let threads = std::env::var("XSCAN_PR_THREADS")
        .ok()
        .and_then(|v| v.parse::<usize>().ok())
        .map(|t| t + 1)
        .unwrap_or_else(|| std::thread::available_parallelism().map(|n| n.get()).unwrap_or(4));
    let pr = PrParams {
        depth: depth as i32,
        full_depth: 5,
        samples_per_node: prm.samples_per_node,
        point_weight: prm.point_weight,
        scale: 1.1,
        iters: 8,
        threads: threads.saturating_sub(1).max(1) as i32,
        linear_fit: 0,
    };
    let raw = poisson::reconstruct(&pf, &nf, &pr)?;
    drop(pf);
    drop(nf);
    if cancelled() {
        return Err("cancelled".into());
    }
    progress(MeshStage::Cleaning);
    let (verts, tris) = trim_to_samples(raw.verts, raw.tris, &tree, s * prm.trim);
    let (mut verts, tris) = drop_crumbs(verts, tris);
    laplacian_smooth(&mut verts, &tris, prm.smooth_iters);
    let normals = vertex_normals(&verts, &tris);
    info.verts = verts.len();
    info.tris = tris.len();
    info.seconds = sw.secs();
    if tris.is_empty() {
        return Err("the reconstruction produced no surface".into());
    }
    Ok((Mesh { verts, normals, tris }, info))
}
